src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/ProcessHelper.cs

```diff
@@ -15,21 +17,98 @@ namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions;
 
 public partial class ProcessHelper : IProcessHelper
 {
+    [DllImport("kernel32.dll", SetLastError = true)]
+    private static extern bool IsWow64Process2(IntPtr process, out ushort processMachine, out ushort nativeMachine);
+
+    private const ushort IMAGE_FILE_MACHINE_ARM64 = 0xAA64;
+    private const ushort IMAGE_FILE_MACHINE_UNKNOWN = 0;
+
     /// <inheritdoc/>
     public string GetCurrentProcessLocation()
-    {
-        return Path.GetDirectoryName(GetCurrentProcessFileName());
-    }
+        => Path.GetDirectoryName(GetCurrentProcessFileName());
 
+    /// <inheritdoc/>
     public IntPtr GetProcessHandle(int processId)
-    {
-        return Process.GetProcessById(processId).Handle;
-    }
+        => Process.GetProcessById(processId).Handle;
 
     /// <inheritdoc/>
     public PlatformArchitecture GetCurrentProcessArchitecture()
+        =>
+        IntPtr.Size == 8
+        ? IsArm64()
+            ? PlatformArchitecture.ARM64
+            : PlatformArchitecture.X64
+        : PlatformArchitecture.X86;
+
+    private static bool IsArm64()
     {
-        return IntPtr.Size == 8 ? PlatformArchitecture.X64 : PlatformArchitecture.X86;
+        try
+        {
+            var currentProcess = Process.GetCurrentProcess();
+            if (!IsWow64Process2(currentProcess.Handle, out ushort processMachine, out ushort nativeMachine))
+                throw new Win32Exception();
+
+            // If processMachine is IMAGE_FILE_MACHINE_UNKNOWN mean that we're not running using WOW64 x86 emulation.
+            // If nativeMachine is IMAGE_FILE_MACHINE_ARM64 mean that we're running on ARM64 architecture device.
+            if (processMachine == IMAGE_FILE_MACHINE_UNKNOWN && nativeMachine == IMAGE_FILE_MACHINE_ARM64)
+            {
+                // To distinguish between ARM64 and x64 emulated on ARM64 we check the PE header of the current running executable.
+                return IsArm64Executable(currentProcess.MainModule.FileName);
+            }
+        }
+        catch (Exception)
+        {
+            // We should have Win32Exception() or IOException() but we don't want to block and we'll fallback to emulated path.
+        }
+
+        return false;
+    }
+
+    private static bool IsArm64Executable(string path)
+    {
+        // This document specifies the structure of executable (image) files
+        // https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#general-concepts
+        using Stream fs = new FileStream(path, FileMode.Open, FileAccess.Read);
+        {
```

Comment:
I don't think this inner block is needed.

Reply:
done